sympy/tensor/array/array_comprehension.py

```diff
@@ -253,27 +253,26 @@ def doit(self):
         if not self.is_shape_numeric:
             return self
 
-        return ImmutableDenseNDimArray(self._expand_array())
+        return self._expand_array()
 
     def _expand_array(self):
-        # To perform a subs at every element of the array.
-        def _array_subs(arr, var, val):
-            arr = MutableDenseNDimArray(arr)
-            for i in range(len(arr)):
-                index = arr._get_tuple_index(i)
-                arr[index] = arr[index].subs(var, val)
-            return arr.tolist()
-
-        list_gen = self.function
-        for var, inf, sup in reversed(self._limits):
-            list_expr = list_gen
-            list_gen = []
-            for val in range(inf, sup+1):
-                if not isinstance(list_expr, Iterable):
-                    list_gen.append(list_expr.subs(var, val))
-                else:
-                    list_gen.append(_array_subs(list_expr, var, val))
-        return list_gen
+        res = []
```

Comment:
You can use a `res = MutableDenseNDimArray.zeros(shape)` object. At the end, call `res.as_immutable()` to make it immutable.

Reply:
Yes but I thought that using the `MutableDenseNDimArray` will make the code more complicated. We have to calculate the index for every value. In terms of efficiency, wouldn't this current method be better?